module/zfs/zfs_ioctl.c

```diff
@@ -2594,6 +2594,44 @@ zfs_prop_set_special(const char *dsname, zprop_source_t source,
 		}
 		break;
 	}
+	case ZFS_PROP_LONGNAME:
+	{
+		zfsvfs_t *zfsvfs;
+		dsl_dataset_t *ds;
+
+		/*
+		 * Ignore the checks if the property is being applied as part of
+		 * 'zfs receive'. Because, we already check if the local pool
+		 * has SPA_FEATURE_LONGNAME enabled in dmu_recv_begin_check().
+		 */
+		if (source == ZPROP_SRC_RECEIVED) {
+			cmn_err(CE_NOTE, "Skipping ZFS_PROP_LONGNAME checks "
+			    "for dsname=%s\n", dsname);
+			err = -1;
+			break;
+		}
```

Comment:
Do we really need this special case?  Received stream may be inconsistent/malicious.

Reply:
Right, I'll remove it.

Reply:
Actually, I had to add this back because when recving a new dataset, zfsvfs_hold would fail.
Maybe we shouldn't call zfsvfs_hold here, but I'll just restore to the original code.
I don't think there would be problem with malformed stream anyway because as the comment says we already checked longname feature, so there's no problem with enable the property on dataset.